keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -259,66 +239,76 @@ the API:
   forwarding requests or persisting data) cannot be undone, but no future
   actions should be allowed.
 
-The examples below include information about what happens when the ReferenceGrant
-is removed as data points.
-
 ### Example Usage
 
-#### Gateway API Gateway Referencing Secret
+#### Gateway API Cross-Namespace Secret Reference
 
-In this example (from the Gateway API docs), we have a Gateway in the
-`gateway-api-example-ns1` namespace, referencing a Secret in the
-`gateway-api-example-ns2` namespace. The following ReferenceGrant allows this:
+Authors of APIs can bundle ClusterReferencePatterns with their API definitions
+for any references they expect to be common. For example, Gateway API could
+bundle the following ClusterReferencePattern to describe how Gateways reference
+`kubernetes.io/tls` Secrets. Note that this also allows API authors to define a
+`SameNamespace` `baselineGrant` that would automatically authorize all
+references following this pattern without crossing namespace boundaries.
 
 ```yaml
-apiVersion: gateway.networking.k8s.io/v1beta1
-kind: Gateway
+kind: ClusterReferencePattern
+apiVersion: reference.authorization.k8s.io/v1alpha1
 metadata:
-  name: cross-namespace-tls-gateway
-  namespace: gateway-api-example-ns1
-spec:
-  gatewayClassName: acme-lb
-  listeners:
-  - name: https
-    protocol: HTTPS
-    port: 443
-    hostname: "*.example.com"
-    tls:
-      # There's a Kind/Resource mismatch here, which sucks, but it is not
-      # easily fixable, since Gateway is already a beta, close to GA
-      # object.
-      certificateRefs:
-      - kind: Secret
-        group: ""
-        name: wildcard-example-com-cert
-        namespace: gateway-api-example-ns2
----
-apiVersion: authorization.k8s.io/v1alpha1
+  name: gateway-tls
+group: gateway.networking.k8s.io
+resource: gateways
+version: v1
+path: ".spec.listeners[*].tls.certificateRefs[*]"
```

Comment:
What is the schema contained in `certificateRefs`?  Name, namespace,name, group/resource/namespace/name?  Apologies if this is somewhere outside the diff

Reply:
This is actually a big open question with the updated design. Currently we specify that each reference needs to be `To` a specific group and resource to be authorized by a (Cluster)ReferenceGrant. One potential interpretation of that is that we would just take the name(s) at this JSON path and assume they were to the group and resource specified by the (Cluster)ReferenceGrant. That does not feel particularly safe though, so another option would be to ensure that the Group and Resource specified by the (Cluster)ReferenceGrant also aligned with the reference here.

Reply:
cc @liggitt @enj @stealthybox @luxas

Reply:
We can implement a sort of duck-typing behavior where is the `path` is:
- a string representing name in the same namespace
- an object that has a name field but no namespace field (assume same namespace)
- an object that has a name and namespace field (if empty, assume same namespace)

This feels like a decent way to support
- Ingress API's use of `secretName`
- Pod style `secretRef` and `envFrom` API's
- API's that have cross-object references

It also satisfies Flux's current API's which don't specify a group for cross-namespace refs -- just and object with name, namespace, version and kind.

Another sensible thing to do would be to also attempt to parse any `group`/`resource`/`kind` present in the object at the path.

Since every grant, whether it's cluster-level or namespaced, specifies a from+to+for, it will usually be unambiguous whether a reference to a specific resource for a specific purpose is allowed.
The edge case is if an API makes a reference where the resource group or kind it's operating on can be changed in a different part of the API spec. Then potentially, you could have ClusterReferenceGrants listing and granting names for group+version+resource that isn't intended. (This isn't much of an issue with namespaced ReferenceGrants because they list specific names already)

One way to support this style of strange API would be to add some sort of compound condition to the ClusterReferenceGrant `path`  that lets you say "and this other field must equal this String".
ex: `config.kind == ConfigMap` or `sourceRef.kind == GitRepository`
This would allow API owners to make more specific ClusterReferenceGrants.
It also opens a use-case for things like `gatewayClass == Contour`. (whether we want to encourage that is another question)
This might already be possible if we use `path` with some sort of compound CEL expression?

The main goal of this very complex heuristic behavior is to support many different types of API's.
The alternative would be to add fields to ClusterReferenceGrant that tell it what type it should expect at the path and how to parse or assume the needed information:
- name
- namespace
- group
- resource / kind

Reply:
@stealthybox: sounds like good grounds for a thing to mark **unresolved**, and a couple of entries in this KEP's Alternatives.

I hope that we can make this easy enough to teach for the common case of “I want to grant access to a thing that I name (eg Ingress / `secretName`)”, and still leave room during beta to make the more complex cases possible.

I also recommend sketching out the docs for how we explain this, because if that proves difficult, maybe the API isn't right yet.